### ISSUE TITLE
Filtering possibly uses the wrong indexes

### ISSUE THREAD

**Issue by AMBudnik** [handsontable/handsontable#4442]
### Description

Filtering can be provided with incorrect indexes. 
Possibly broken by v`1.11.0-beta1`

### Steps to reproduce

**CASE 1**
In this example, we use moving rows feature.
1. Log data via button > result `0,0` 
2. Move the first row to the second position
3. Log data via button > result `0,0` 
4. Use `Filter_by_value` to uncheck `A1`
5. Log data via button > result `0, null` 

+ the `A1` row is still visible but `A2` is gone.

**CASE 2**
The same situation happens by sorting
1. Log data via button > result `0,0` 
2. Sort data descending (`A1` will be the last one)
3. Log data via button > result `0,9`
4. Use `Filter_by_value` to uncheck `A1`
5. Log data via button > result `0, 6`

The row has been filtered but is has an index of `6`. 

**CASE 3**
When we combine moving rows and sorting the result is also unexpected but different.
1. Move the first row to the second position
2. Sort data descending (`A1` will be the last one)
3. Use `Filter_by_value` to uncheck `A1`

In a result the `A1` is still visible, `A2` is gone and we have an extra row at the end of the table.

### Demo

https://jsfiddle.net/JoshAdams/vn4rxLsm/

### Your environment
* Handsontable version: 1.13.0/0.34.0
* Browser Name and version: Chrome 60
* Operating System: Windows 10

`[Gr #7594]`

**Comment by wszymanski** [comment]
In the core of one of the problems is work of `trimRows` plugin. It should be fixed within internal ticket system (GL #186).

**Comment by AMBudnik** [comment]
Issue stays the same for v1.15.0

**Comment by mountaineerfan5** [comment]
I am not sure if this coincides with this problem, but I have experienced some problems with the indexing of trimRows. In the example I am trimming the first two rows of the data. If you insert a row above the 1st visual row, which is actually the 3rd row for the physical data. It seems like it puts it above the 1st row of the physical data, so now my trimRows has moved down. I want to keep my trimRows at the 0th and 1st index of the physical data. I'm not sure if this is how trimRows was intended to work, but the behavior seems a bit odd. 

Here is a jsFiddle: http://jsfiddle.net/mountaineerfan5/L8mwwqqm/3/
1.  Use the context menu to insert a row above the first visual row in the handsontable
2.  Click the get data button to dump the data to the console

The trimRows is now moved down to the 1st and 2nd index in the handsontable.

Handsontable version: 1.15.1
Browser Name and version: Chrome 63
Operating System: Windows 10

Thanks.

**Comment by AMBudnik** [comment]
Hi @mountaineerfan5 

the `trimRows` uses physical indexes. I have added the `hot.render()` at line `23` to rerender the view and it seems to work fine. Here's the example http://jsfiddle.net/awt4hgjz/

![gif](https://user-images.githubusercontent.com/10757813/35962814-1416f310-0cb3-11e8-89a5-f677a8fa0d0a.gif)

**Comment by mountaineerfan5** [comment]
Sorry I don't think I explained what I meant well. After rendering the data in my previous example the table visually looks fine, but the inserted row should have been the 3rd index for the physical array but it was inserted as the 0th index. I think I have a better example to show my problem.

My main goal in this, is to get the data that is in the handsontable and set it to a variable including the trimmed row. But when I look at hot.getSourceData() or hot.getData() after I have untrimmed the rows and rendered the table row indexes aren't correct. 

JsFiddle: http://jsfiddle.net/mountaineerfan5/L8mwwqqm/6/

Now this time:
1. Use the context menu on the 1st row but this time click insert row below
2. Click the button. This will untrim the rows, render the data, and dump the table's data to the console.

When you untrim the rows and render the handsontable then the row is put in between the two trimmed rows.  After untrimming the rows it should be the 4th row index but it stays put as the 2nd row of the table. You can also see this in the data dumped to the console. Whether it is the visual data (hot.getData()) or if it is the physical data (hot.getSourceData()).

I hope this helps clarify what I was trying to demonstrate.
Thanks.

**Comment by AMBudnik** [comment]
Hm... I am wondering how we can make it work as you want. Cause basically we have only two options to add a row, by using a visual or physical index. This demo shows the problem http://jsfiddle.net/8L8awbvx/
Even when we use the physical index which works well when we untrim the data, the empty rows does not show in the first place where you wanted it to be.

**Comment by mountaineerfan5** [comment]
Yeah that's exactly what I was talking about. We decided to go away from trimRows, it didn't quite work how we wanted it too. If it could insert the rows at the right index then we would use it, but instead we have to do a work around. We wanted to carry the trimRows with a column, because they were attributes for the columns. Instead of trimRows we JSON encoded each cell of the first row and made the renderer display the information for our users. We also had to adjust some functions like beforeChange so whatever the user types in we adjust it back into a JSON format. We changed afterBeginEditing cause the textarea would think that it needs to put in the whole JSON string into it. We had to change other functions too, but here's some sample code of what we did.
     //Renderer for first row;
    firstRowRenderer: function(instance, td, row, col, prop, value, cellProperties) {
        Handsontable.renderers.TextRenderer.apply(this, arguments);
        if (value !== null) {
            var temp = JSON.parse(value);
            td.innerHTML = temp.name;
        }
    }

beforeChange: function (changes, source) {
        //If changing the variable names, replace the variable_names prop inside the JSON object
        for (var i = 0, length = changes.length; i < length; i++) {
            if (changes[i][0] === 0) {
                // check if null; Happens on a newly created cell that has never held a value
                // ex: creating a new column
                if (changes[i][2] !== null) {
                    var old_val = JSON.parse(changes[0][2]);
                    var new_val = changes[0][3];
                    old_val.name= new_val;
                    changes[0][3] = JSON.stringify(old_val);
                }
            }
        }
        return changes;
    },
    afterBeginEditing: function(row, col) {
        // extract the variable_names prop of the JSON object as the val of the text input
        if (row === 0) {
            var elem = document.querySelector('textarea.handsontableInput');
            var temp = JSON.parse(hot_edit.getDataAtCell(row, col));
            var name = temp.names;
            var size = name.length * 10;
            var col_width = edit_data_controller.col_widths[col];

            if (size < col_width) {
                size = col_width;
            }

            elem.value = name;
            elem.style.width = size + 'px';
        }
    }

We also had to add code for newly created columns to have the JSON object in the first row and adjust code in other places. It would've been easier if trimRows worked like we needed it too, but I guess we found a work around that works so far.

**Comment by mountaineerfan5** [comment]
Yeah that's exactly what I was talking about. We decided to go away from trimRows, it didn't quite work how we wanted it too. If it could insert the rows at the right index then we would definitely be using it, but instead we have to do a work around. We wanted to carry the trimRows with a column, because they were attributes for the columns. Instead of trimRows we JSON encoded each cell of the first row (i.e. {names: "text to display", attr1: "1st attr", attr2: "2nd attr", ...}  and made the renderer display the information for our users. We also had to adjust some functions like beforeChange so whatever the user types in we adjust it back into a JSON format. We changed afterBeginEditing cause the textarea would think that it needs to put in the whole JSON string into it. We had to change other functions too, but here's some sample code of what we did.
```
//Renderer for first row;
    firstRowRenderer: function(instance, td, row, col, prop, value, cellProperties) {
        Handsontable.renderers.TextRenderer.apply(this, arguments);
        if (value !== null) {
            var temp = JSON.parse(value);
            td.innerHTML = temp.name;
        }
    }

beforeChange: function (changes, source) {
        //replace the names prop inside the JSON object if row == 0
        for (var i = 0, length = changes.length; i < length; i++) {
            if (changes[i][0] === 0) {
                // check if null; Happens on a newly created cell that has never held a value
                // ex: creating a new column
                if (changes[i][2] !== null) {
                    var old_val = JSON.parse(changes[0][2]);
                    var new_val = changes[0][3];
                    old_val.name= new_val;
                    changes[0][3] = JSON.stringify(old_val);
                }
            }
        }
        return changes;
    },
    afterBeginEditing: function(row, col) {
        // extract the names prop of the JSON object as the val of the text input
        if (row === 0) {
            var elem = document.querySelector('textarea.handsontableInput');
            var temp = JSON.parse(hot_edit.getDataAtCell(row, col));
            var name = temp.names;
            var size = name.length * 10;
            var col_width = js_controller.col_widths[col];

            if (size < col_width) {
                size = col_width;
            }

            elem.value = name;
            elem.style.width = size + 'px';
        }
    }
```

We also had to add code for newly created columns to have the JSON object in the first row and adjust code in other places. It would've been easier if trimRows worked like we needed it too, but I guess we found a work around that works so far.

**Comment by AMBudnik** [comment]
Wow. That's a smart move. Sounds pretty solid.
Sadly we do not have anything that will support this kind of operations out of the box

**Comment by AMBudnik** [comment]
Tested with version 0.38.1

**Case 1**
Still, present but the step 3.  return `0,1` instead of `0,0`

**Case 2**
The same as before

**Case 3**
The same as before

**Comment by AMBudnik** [comment]
Tests for version `8.0.0-rc1` for all the 3 cases succeded.

### Case 1 - **Solved**
The indexes are counted correctly
```
physicalrow = 0
(index):52 logicalrow = 0
(index):51 physicalrow = 0
(index):52 logicalrow = 1
(index):51 physicalrow = 0
(index):52 logicalrow = null
```
The data is not missing
![image](https://user-images.githubusercontent.com/10757813/64961505-f265df00-d895-11e9-951f-afe59f476fd0.png)

### Case 2 - **Solved**
The filtered out index returns `null` instead of `6`.
The correct row has been filtered out.
![image](https://user-images.githubusercontent.com/10757813/64961617-35c04d80-d896-11e9-99c3-4fc6b39c68ee.png)

### Case 3 - **Solved**
Previously
> In a result the A1 is still visible, A2 is gone and we have an extra row at the end of the table.

Currently `A1` is correctly filtered out and `A2` is visible. There are no extra rows.

**Comment by AMBudnik** [comment]
In the official 8.0.0-beta1 https://jsfiddle.net/zksy4h2m/ 

- [x] case1
- [x] case2
- [x] case3

are solved! Well done.

**Comment by AMBudnik** [comment]
All issues remained fixed for 8.0.0-beta2-rev12 https://jsfiddle.net/AMBudnik/zkwbt2a8/

**Comment by AMBudnik** [comment]
Demo https://jsfiddle.net/hmp5a74n/
 * [x] Case 1 - solved
 * [x] Case 2 - solved
 * [x] Case 3 - solved
